File: client/src/discord_bridge/bot.py
"""Discord bot setup and main event loop.

Pure forwarder — no LLM. Routes messages between Discord and the relay.
"""

from __future__ import annotations

import asyncio

import discord
import structlog

from ..config import Config
from ..relay.client import RelayClient
from .commands import CommandHandler
from .router import MessageRouter

log = structlog.get_logger()


async def start_discord_bot(config: Config, relay: RelayClient):
    """Start the Discord bot. Runs until cancelled."""
    intents = discord.Intents.default()
    intents.message_content = True
    intents.members = True

    bot = discord.Client(intents=intents)
    command_handler = CommandHandler(config, relay)
    router = MessageRouter(config, relay)

    @bot.event
    async def on_ready():
        log.info("discord.ready", user=str(bot.user), guilds=len(bot.guilds))
        # Start outbound relay → Discord forwarding loop
        asyncio.create_task(router.start_outbound_loop(bot))

    @bot.event
    async def on_message(message: discord.Message):
        # Ignore bot's own messages
        if message.author == bot.user:
            return

        # Ignore DMs for now
        if not message.guild:
            return

        # Check if it's a command (starts with /)
        if message.content.startswith("/"):
            await command_handler.handle(message)
            return

        # Check if bot is mentioned
        if bot.user in message.mentions:
            # Strip mention and treat rest as command
            content = message.content.replace(f"<@{bot.user.id}>", "").strip()
            if content:
                # Try to parse as "agentname message"
                parts = content.split(None, 1)
                if len(parts) >= 2:
                    await command_handler.handle_raw(message, parts[0], parts[1])
                elif len(parts) == 1:
                    await command_handler.handle_raw(message, parts[0], "")

    try:
        await bot.start(config.discord.token)
    except discord.LoginFailure:
        log.error("discord.login_failed", hint="Check DISCORD_TOKEN in .env")
    except asyncio.CancelledError:
        await bot.close()

File: internal/discord/bot.go
// Package discord implements the WRAI.TH Discord bridge.
// Pure message forwarder — zero LLM tokens. Station mode only.
package discord

import (
	"log"

	"github.com/bwmarrin/discordgo"

	"agent-relay/internal/client"
)

// Bot wraps the Discord session and wires commands + routing.
type Bot struct {
	session  *discordgo.Session
	config   *client.Config
	relay    *client.RelayClient
	commands *CommandHandler
	router   *MessageRouter
}

// NewBot creates a Discord bot from config.
func NewBot(cfg *client.Config, relay *client.RelayClient, sse *client.SSEClient) (*Bot, error) {
	dg, err := discordgo.New("Bot " + cfg.Discord.Token)
	if err != nil {
		return nil, err
	}

	dg.Identify.Intents = discordgo.IntentsGuildMessages | discordgo.IntentsMessageContent

	b := &Bot{
		session:  dg,
		config:   cfg,
		relay:    relay,
		commands: NewCommandHandler(cfg, relay),
		router:   NewMessageRouter(cfg, relay, sse),
	}

	dg.AddHandler(b.onMessageCreate)
	dg.AddHandler(b.onReady)

	return b, nil
}

// Start connects to Discord and begins routing.
func (b *Bot) Start() error {
	if err := b.session.Open(); err != nil {
		return err
	}

	// Start outbound routing (SSE-driven, not polling)
	b.router.Start(b.session)

	log.Printf("[discord] connected as %s", b.session.State.User.Username)
	return nil
}

// Stop disconnects from Discord.
func (b *Bot) Stop() {
	b.router.Stop()
	b.session.Close()
	log.Printf("[discord] disconnected")
}

func (b *Bot) onReady(s *discordgo.Session, r *discordgo.Ready) {
	log.Printf("[discord] ready: %s#%s (guilds: %d)", r.User.Username, r.User.Discriminator, len(r.Guilds))
}

func (b *Bot) onMessageCreate(s *discordgo.Session, m *discordgo.MessageCreate) {
	// Ignore own messages
	if m.Author.ID == s.State.User.ID {
		return
	}

	// Ignore messages not starting with /
	if len(m.Content) == 0 || m.Content[0] != '/' {
		return
	}

	b.commands.Handle(s, m)
}
